src/Compiler/Checking/CheckExpressions.fs

```diff
@@ -5335,100 +5335,107 @@ and CheckForAdjacentListExpression (cenv: cenv) synExpr hpa isInfix delayed (arg
 /// method applications and other item-based syntax.
 and TcExprThen (cenv: cenv) overallTy env tpenv isArg synExpr delayed =
     let g = cenv.g
+    
+    match env.eCachedImplicitYieldExpressions |> List.tryPick (fun (se, ty, e) -> if System.Object.ReferenceEquals(se, synExpr) then Some (ty, e) else None) with
```

Comment:
`TcExprThen` is very common.
Could the list search be prepended with a lookup, e.g. by range of the synexpr? And then only scan a list from a multimap for the same range.

The compiler codebase has an immutable MultiMap as well as a `HashMultiMap`

Reply:
Hey, I couldn't use Map/MultiMap (since range is not comparable) so I used HashMultiMap and cleaned the entry after the inner type-check. Hope that's what you ment

Reply:
Yes, this is what I meant - to reduce the overhead for all other method applications.